### ISSUE TITLE
Improving check_model() for GLMs

### ISSUE THREAD

**Issue by bwiernik** [easystats/performance#376]
The current selection of plots returned by `check_model()` for GLMs aren't ideal in a few ways.

1. They are missing a linearity check (fitted vs residuals). For binomial models, this should be a called to `binned_residuals()`. For other families, the standard check is fine.
2. For binomial models, the constant variance plot should be omitted.
3. For binomial models, the residual QQ plot is hard to interpret.
3. For non-bernoulli models, we should include a plot for checking overdispersion.

For the latter few points, the DHARMa package provides an easy-to-interpret approach for checking distributional assumptions from qq plots and problems with fitted vs residual plots using quantile residuals. We might consider soft-importing DHARMa or re-implementing those approaches.
https://cran.r-project.org/web/packages/DHARMa/vignettes/DHARMa.html
https://github.com/florianhartig/DHARMa/issues/33

**Comment by bbolker** [comment]
One standard check for bernoulli models is to do a binned binomial proportions plot (this is, loosely, a graphical equivalent of the Hosmer-Lemeshow test, and is implemented in `arm::binnedplot()`

**Comment by bwiernik** [comment]
Yep, that's implemented in performance as `binned_residuals()`

**Comment by bwiernik** [comment]
No worries! Thanks for your comment!

**Comment by DominiqueMakowski** [comment]
Should we incorporate the plot into check_model?

**Comment by bwiernik** [comment]
I think we also might want to add a check_predictions() plot to check_model() for all models, perhaps dropping the normal density plot?

**Comment by bbolker** [comment]
This isn't the place, but I may have some Opinions about `binned_model` (I would like a smoothed line added, binomial CIs on the points; I would rather have binomial CIs on the individual points and a confidence interval on the smooth rather than have a dichotomizing region that indicates 'good' and 'bad' points ...). I see that this is copying what `arm::binnedplot` does though ...

**Comment by bwiernik** [comment]
I like that. Have an example?

**Comment by bbolker** [comment]
Not yet but could do one tomorrow.

**Comment by bwiernik** [comment]
That would be great!

**Comment by jebyrnes** [comment]
Would it be easier to pull in https://github.com/florianhartig/DHARMa as its very flexible in terms of model type?

**Comment by bbolker** [comment]
https://gist.github.com/bbolker/945e684dc148239364ab0f99c9a1beed

I realized that the residuals from binomials are not binomial (so constructing binomial CIs isn't useful); however I think that a binned plot using `mean_cl_boot()` to generate CIs is probably fine.  See examples and decide what you think is useful or not ...

**Comment by bwiernik** [comment]
I think it would be good to put `check_predictions()` (pp_check) into the this plot array. For linear models, we have two normality checking plots. I think it would be good to swap out the density plot for a posterior predictions plot. Thoughts?

**Comment by strengejacke** [comment]
> For binomial models, the constant variance plot should be omitted.

Which panel do you exactly mean? I think internally, the names are different from plot titles. I assume you mean the plot in the top right panel?

![image](https://user-images.githubusercontent.com/26301769/155725206-93440256-5fae-42b4-8ac2-3609ca05894d.png)

**Comment by bwiernik** [comment]
Yes, the top right one.

**Comment by strengejacke** [comment]
Current state:

``` r
library(performance)
model <- glm(vs ~ wt + mpg, data = mtcars, family = "binomial")
check_model(model)
```

![](https://i.imgur.com/GroOqKg.png)

<sup>Created on 2022-02-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
> I think we also might want to add a check_predictions() plot to check_model() for all models, perhaps dropping the normal density plot?

Only the predictions, not the range, right? (i.e. not what would be returned when `check_predictions(model, check_range = TRUE)`, because that would be too much for such little space?)

**Comment by strengejacke** [comment]
@strengejacke TODO: use same colours for binned residuals as in other plots (e.g. blue/red like in VIF)

**Comment by jebyrnes** [comment]
What about using DHARMa and it’s quartile residuals approach? The qqunif plot that results is very intuitive if you are used to qqplots.

**Comment by strengejacke** [comment]
> What about using DHARMa and it’s quartile residuals approach? The qqunif plot that results is very intuitive if you are used to qqplots.

Do you mean for binomial models?

**Comment by jebyrnes** [comment]
Yes? Or any glm. Also meant quantile not quartile. Stupid autocorrect. Ha!

**Comment by strengejacke** [comment]
``` r
library(performance)
model <- glm(vs ~ wt + mpg, data = mtcars, family = "binomial")
check_model(model)
```

![](https://i.imgur.com/iBkBRWI.png)

``` r
m <- lm(mpg ~ wt + cyl + gear + disp, data = mtcars)
check_model(m)
```

![](https://i.imgur.com/KCU3rQZ.png)

<sup>Created on 2022-02-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
@bwiernik should we change the labs for the pp-check panel? First, y and yrep don't sound clear to me. Second, if colours are changed, lines are probably no longer blue and grey. And there's no y and x axis title, leading to empty spaces (which is, however, aligned with the other plots).

**Comment by bwiernik** [comment]
Yeah, I can clarify the labels to match the other plots style

**Comment by strengejacke** [comment]
> For non-bernoulli models, we should include a plot for checking overdispersion.

We don't have such a plot right now...

**Comment by strengejacke** [comment]
I just saw that `binned_residuals()` includes SEs, maybe we can use those for constructing CIs? It would probably good to compare against `mean_ci_boot`, but I'm not sure how...

``` r
library(ggplot2)
library(performance)
model <- glm(vs ~ wt + mpg, data = mtcars, family = "binomial")
plot(result) + geom_errorbar(aes(ymin = CI_low, ymax = CI_high, colour = group))
#> Error in plot(result): object 'result' not found

aa <- broom::augment(model)

gg1 <- ggplot(aa, aes(.fitted, .resid)) +
  stat_summary_bin(fun.data = mean_cl_boot) +
  geom_hline(yintercept = 0, lty = 2) +
  geom_smooth()

gg1
#> `geom_smooth()` using method = 'loess' and formula 'y ~ x'
#> Warning: Removed 11 rows containing missing values (geom_segment).
```

![](https://i.imgur.com/GlfEn8N.png)

<sup>Created on 2022-02-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by bwiernik** [comment]
> We don't have such a plot right now... 

For future improvement ;)

**Comment by bwiernik** [comment]
Thoughts on the above 3 plots? @strengejacke @DominiqueMakowski @IndrajeetPatil @mattansb @vincentarelbundock

**Comment by strengejacke** [comment]
For me, the first looks most intuitive

**Comment by bbolker** [comment]
You didn't ask me, but: 

1. I think this kind of plot is most useful for diagnosing/eye-balling the variance-mean relationship (e.g. do the raw/un-pearsonized residuals most resemble Poisson-type (var ~ mean), NB1 (var = c*mean, c>1), NB2 (var = mu*(1+mu/k), k>0), Gamma/log-normal (var = c*mean^2), ... ?
2. (a) The left-hand plot here puzzles me; not sure what's on the axes?  (b) I actually like this one best (in perfect contrast to @jebyrnes), if the multiple lines are explained. In my experience people usually use the scale-location plot only to detect *trends* in scale, but there's no reason it couldn't be calibrated to set the *expected* scale/y-axis value to 1 ...
3. I don't think the Q-Q plot is typically useful/used for diagnosing overdispersion. DHARMa does a **test** for overdispersion (I don't remember how) and reports it on the Q-Q panel, but normally the Q-Q plot is only used for agreement in shape (i.e., we only look to see if the pattern is linear, rather than comparing the slope to an expected theoretical slope &mdash; this is in parallel with the point above that scale-location plots are usually used to compare pattern, not magnitude ...

**Comment by bwiernik** [comment]
The qqplots can be useful for overdispersion because that shape with fat tails is a characteristic pattern of overdispersion. That said, I'm really not a fan of qqplots in general, and I agree that other plots like posterior predictive densities and fitted-residuals plots are probably more useful

**Comment by mattansb** [comment]
I also like the first one (:

I don't think that (most people) can diagnose dispersion by looking a a qq-plot, but maybe that's just me (:

**Comment by strengejacke** [comment]
@bwiernik How can we create the first plot variance versus expected)? We may include this in the forthcoming update.

**Comment by bwiernik** [comment]
tl;dr I think we should go with the second plot below. What do you think?


Okay, so I looked into what SAS was doing with that first plot. Super disappointing. It's actually just plotting the model-predicted mean agains the model-predicted variance (so for a Poisson model, always a straight line). So there's no actual observed data there, only predictions. I wouldn't even call it a diagnostic plot -- you need to have already fit a model with over dispersion before it will confirm that, yes, there was over dispersion that needed to handled. 

So I played around and came up with two ideas

![image](https://user-images.githubusercontent.com/4773225/158478485-fc91a64c-782d-4b52-8548-12f767bd089a.png)

In the top plot, we show the standardized residuals. These should follow a normal distribution, so most points should be within ±2 and almost no points should be outside ±4. There is a pretty big number larger that +2.

The second plot takes the squared residuals and fits a `geom_smooth()` to them. This gives us the observed average squared residual (i.e., variance) for each Predicted value. It then shows the comparison against the model-predicted variance. In this plot, we see that the observed variance is much higher than expected at a bunch of ranges.

I think the second plot is probably better. It doesn't have individual points, but it is still based on the actual observed residuals, and I think it more clearly indicates the violation of the assumption.

What do you think?


Here is an example of the above data with a model better handling the dispersion (negative binomial, zero-inflated, with a variable dispersion model):

![image](https://user-images.githubusercontent.com/4773225/158478545-737d6933-6827-4a70-965d-901ab4c4816e.png)

**Comment by bwiernik** [comment]
@bbolker Either of the above require computation of Pearson residuals and/or model-expected conditional variance. Do you think that a function could be added for the conditional variance in glmmTMB for the various count families, including with zero-inflation and variable dispersion (e.g., for nbinom2, `V = Predicted * (1 + Predicted / Disp) * (1 - Prob) * (1 + Predicted * (1 + Predicted / Disp) * Prob)`)?

**Comment by strengejacke** [comment]
2nd plot looks great, and is easy to understand even for non-experts. Can you share some code, so I can implement it in _performance_ (and _see_), or do you want to do that?

**Comment by bwiernik** [comment]
I'll through some code up

**Comment by bwiernik** [comment]
[docvisit.txt](https://github.com/easystats/performance/files/8262955/docvisit.txt)


```r
docvisit <- readr::read_table(file.path("docvisit.txt"))

docvisit

library(glmmTMB)

mp <- glmmTMB(
  doctorco ~ sex + illness + income + hscore, 
  data = docvisit,
  family = poisson()
)

mnb <- glmmTMB(
  doctorco ~ sex + illness + income + hscore, 
  data = docvisit,
  family = nbinom2()
)

mzip <- glmmTMB(
  doctorco ~ sex + illness + income + hscore, 
  ziformula = ~ age,
  data = docvisit,
  family = poisson()
)

mzinb <- glmmTMB(
  doctorco ~ sex + illness + income + hscore,
  ziformula = ~ age,
  data = docvisit,
  family = nbinom2()
)

mzinbd <- glmmTMB(
  doctorco ~ sex + illness + income + hscore + age,
  ziformula = ~ sex + illness + income + hscore + age,
  dispformula = ~ sex + illness + income + hscore + age,
  data = docvisit,
  family = nbinom2()
)

ep <- modelbased::estimate_expectation(mp) |> 
  dplyr::mutate(
    Residuals = -1 * Residuals,
    Res2 = Residuals^2,
    Disp = 0,
    Prob = 0,
    V = Predicted,
    PredV = predict(lm(Res2 ~ poly(Predicted, 2))),
    StdRes = residuals(mp, type = "pearson")
  )

enb <- modelbased::estimate_expectation(mnb) |> 
  dplyr::mutate(
    Residuals = -1 * Residuals,
    Res2 = Residuals^2,
    Disp = predict(mnb, type = "disp"),
    Prob = 0,
    V = Predicted * (1 + Predicted / Disp),
    PredV = predict(lm(Res2 ~ poly(Predicted, 2))),
    StdRes = residuals(mnb, type = "pearson")
  )

ezip <- modelbased::estimate_expectation(mzip) |> 
  dplyr::mutate(
    Residuals = -1 * Residuals,
    Res2 = Residuals^2,
    Disp = 0
    Prob = predict(mzip, type = "zprob"),
    V = Predicted * (1 - Prob) * (1 + Predicted * Prob),
    PredV = predict(lm(Res2 ~ poly(Predicted, 2))),
    StdRes = Residuals / sqrt(V)
  )

ezinb <- modelbased::estimate_expectation(mzinb) |> 
  dplyr::mutate(
    Residuals = -1 * Residuals,
    Res2 = Residuals^2,
    Disp = predict(mzinb, type = "disp"),
    Prob = predict(mzinb, type = "zprob"),
    V = Predicted * (1 + Predicted / Disp) * (1 - Prob) * (1 + Predicted * (1 + Predicted / Disp) * Prob),
    PredV = predict(lm(Res2 ~ poly(Predicted, 2))),
    StdRes = Residuals / sqrt(V)
  )

ezinbd <- modelbased::estimate_expectation(mzinbd) |> 
  dplyr::mutate(
    Residuals = -1 * Residuals,
    Res2 = Residuals^2,
    Disp = predict(mzinbd, type = "disp"),
    Prob = predict(mzinbd, type = "zprob"),
    V = Predicted * (1 + Predicted / Disp) * (1 - Prob) * (1 + Predicted * (1 + Predicted / Disp) * Prob),
    PredV = predict(lm(Res2 ~ poly(Predicted, 2))),
    StdRes = Residuals / sqrt(V)
  )

p1 <- ggplot(ezinbd) + aes(x = Predicted) + 
  geom_point(aes(y = StdRes)) + 
  geom_hline(
    yintercept = c(-2, 2, -4, 4), 
    linetype = c("solid", "solid", "dashed", "dashed"),
    color = c(rep(see::social_colors("green"), 2), rep(see::social_colors("blue"), 2))
  ) +
  labs(
    title = "Overdispersion and zero-inflation", 
    subtitle = "Most points should be within ±2 (green), few points outside ±4 (blue)",
    x = "Predicted mean",
    y = "Standardized resiuduals"
  ) +
  see::theme_modern(plot.title.space = 2)

p2 <- ggplot(ezinbd) + aes(x = Predicted) + 
  geom_smooth(aes(y = V), color = see::social_colors("blue"), se = FALSE) + 
  geom_smooth(aes(y = Res2), color = see::social_colors("green")) + 
  labs(
    title = "Overdispersion and zero-inflation", 
    subtitle = "Observed residual variance (green) should follow predicted residual variance (blue)",
    x = "Predicted mean",
    y = "Residual variance"
  ) +
  see::theme_modern(plot.title.space = 2)

p1 / p2  
```

**Comment by strengejacke** [comment]
ok, first model types are supported (missing will be added later). Some tweaking necessary, e.g. correct colors:

``` r
library(glmmTMB)
library(performance)
data("Salamanders")
m <- glmmTMB(
  count ~ mined + spp + (1 | site),
  family = poisson,
  data = Salamanders
)
check_model(m)
#> `check_outliers()` does not yet support models of class glmmTMB.
```

![](https://i.imgur.com/OrMud1i.png)

<sup>Created on 2022-03-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by bwiernik** [comment]
Awesome. Can you link to where this is implemented?

**Comment by strengejacke** [comment]
``` r
library(readr)
docvisit <- read_table2("C:/Users/DL/Downloads/docvisit.txt")
#> Warning: `read_table2()` was deprecated in readr 2.0.0.
#> Please use `read_table()` instead.
#> 
#> -- Column specification --------------------------------------------------------
#> cols(
#>   sex = col_double(),
#>   age = col_double(),
#>   agesq = col_double(),
#>   income = col_double(),
#>   levyplus = col_double(),
#>   freepoor = col_double(),
#>   freerepa = col_double(),
#>   illness = col_double(),
#>   actdays = col_double(),
#>   hscore = col_double(),
#>   chcond1 = col_double(),
#>   chcond2 = col_double(),
#>   doctorco = col_double(),
#>   nondocco = col_double(),
#>   hospadmi = col_double(),
#>   hospad = col_double(),
#>   medicine = col_double(),
#>   prescrib = col_double(),
#>   nonpresc = col_double()
#> )

library(glmmTMB)
#> Warning: package 'glmmTMB' was built under R version 4.1.3
library(performance)

mp <- glmmTMB(
  doctorco ~ sex + illness + income + hscore, 
  data = docvisit,
  family = poisson()
)
plot(check_overdispersion(mp))
#> `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'
#> `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'
```

![](https://i.imgur.com/M16bod6.png)

``` r
mnb <- glmmTMB(
  doctorco ~ sex + illness + income + hscore, 
  data = docvisit,
  family = nbinom2()
)
plot(check_overdispersion(mnb))
#> `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'
#> `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'
```

![](https://i.imgur.com/7k9RZev.png)

``` r
mzip <- glmmTMB(
  doctorco ~ sex + illness + income + hscore, 
  ziformula = ~ age,
  data = docvisit,
  family = poisson()
)
plot(check_overdispersion(mzip))
#> Warning: Cannot test for overdispersion, because pearson residuals are not implemented
#>   for models with zero-inflation or variable dispersion.
#>   Only the visual inspection using 'plot(check_overdispersion(model))' is
#>   possible.
#> `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'
#> `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'
```

![](https://i.imgur.com/uDk9teE.png)

``` r
mzinb <- glmmTMB(
  doctorco ~ sex + illness + income + hscore,
  ziformula = ~ age,
  data = docvisit,
  family = nbinom2()
)
plot(check_overdispersion(mzinb))
#> Warning: Cannot test for overdispersion, because pearson residuals are not implemented
#>   for models with zero-inflation or variable dispersion.
#>   Only the visual inspection using 'plot(check_overdispersion(model))' is
#>   possible.
#> `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'
#> `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'
```

![](https://i.imgur.com/lEPlA3x.png)

``` r
mzinbd <- glmmTMB(
  doctorco ~ sex + illness + income + hscore + age,
  ziformula = ~ sex + illness + income + hscore + age,
  dispformula = ~ sex + illness + income + hscore + age,
  data = docvisit,
  family = nbinom2()
)
plot(check_overdispersion(mzinbd))
#> Warning: Cannot test for overdispersion, because pearson residuals are not implemented
#>   for models with zero-inflation or variable dispersion.
#>   Only the visual inspection using 'plot(check_overdispersion(model))' is
#>   possible.
#> `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'
#> `geom_smooth()` using method = 'gam' and formula 'y ~ s(x, bs = "cs")'
```

![](https://i.imgur.com/lNQSE4Y.png)

<sup>Created on 2022-03-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
> Awesome. Can you link to where this is implemented?

This works for `check_model()`, not yet for `plot()` (because it requires the forthcoming update from _see_ on CRAN). I'll commit the latest changes to working PR branches, so you can reproduce the plots above.

**Comment by strengejacke** [comment]
Ok, you need the latest _insight_ from GitHub/r-universe, _performance_ from this PR (https://github.com/easystats/performance/pull/404), and latest _see_ from GitHub / r-universe. Then, the above examples work. If you install _performance_ from master instead of PR, "only" the `check_model()` including overdispersion plots work.

**Comment by strengejacke** [comment]
In future versions, we'll be able to have both plot-types for overdispersion...

``` r
library(glmmTMB)
library(performance)
data("Salamanders")
m <- glmmTMB(
  count ~ mined + spp + (1 | site),
  family = poisson,
  data = Salamanders
)
check_model(m)
```

![](https://i.imgur.com/XFq0VSc.png)

``` r
check_model(m, plot_type = 2)
```

![](https://i.imgur.com/D7m2LuM.png)

**Comment by DominiqueMakowski** [comment]
This design is somewhat a bit at odds with our traditional opinionated API, rather than having different `plot_types`, I'd just pick one version which we think it's the best and stick with it (until the next improvement)... or, if they show different stuff, have the two overdispersion plots side by side in the corner (subplots)?

**Comment by strengejacke** [comment]
Yeah, I think for now we stick to plot 1, the default. I just implemented that option, which is undocumented for now, bnut we could make it in a similar way as with normality plots, where we have 3 different types:
https://easystats.github.io/see/articles/performance.html#check-for-normal-distributed-residuals

**Comment by mattansb** [comment]
> This design is somewhat a bit at odds with our traditional opinionated API, rather than having different plot_types, I'd just pick one version which we think it's the best and stick with it 

Agreed.

**Comment by strengejacke** [comment]
@mccarthy-m-g suggestion looks rather easy to implement.

**Comment by mccarthy-m-g** [comment]
@strengejacke What would a PR for this involve? I could get a draft started if this is the solution you want to go for.

**Comment by bwiernik** [comment]
Let's call the function `check_residuals()`

@mccarthy-m-g You can add the function to a new check_residuals.R file here in the performance package. Take a look at one of the other check_ functions like check_normality.R for an example of the documentation syntax and structure. 

Then open a PR here and we can merge it in. After that, then we can move over to the see package repo and add the plotting function there.

**Comment by mccarthy-m-g** [comment]
Hi all, I just opened a new issue to discuss the implementation for `check_residuals()` (#595). There are a few things that should be resolved before getting a PR started.

**Comment by strengejacke** [comment]
This is the current development stage. We see a mismatch between the tests based on simulated residuals and generated plots for following families/models:

- mnb / nbinom2() (plot suggests underdispersion)
- mzinb / nbinom2() with ZI (plot suggests underdispersion)

``` r
library(performance)
library(glmmTMB)
library(readr)
docvisit <- read_table2("C:/Users/Daniel/Downloads/docvisit.txt")

mp <- glmmTMB(
  doctorco ~ sex + illness + income + hscore, 
  data = docvisit,
  family = poisson()
)
out <- check_overdispersion(mp)
out
#> # Overdispersion test
#> 
#>        dispersion ratio =    1.808
#>   Pearson's Chi-Squared = 9375.539
#>                 p-value =  < 0.001
#> Overdispersion detected.
plot(out)
#> `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'
#> `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'
```

![](https://i.imgur.com/GOkFh1t.png)

``` r

mnb <- glmmTMB(
  doctorco ~ sex + illness + income + hscore, 
  data = docvisit,
  family = nbinom2()
)
out <- check_overdispersion(mnb)
out
#> # Overdispersion test
#> 
#>  dispersion ratio = 1.005
#>           p-value = 0.816
#> No overdispersion detected.
plot(out)
#> `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'
#> `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'
```

![](https://i.imgur.com/CSDeA30.png)

``` r

mzip <- glmmTMB(
  doctorco ~ sex + illness + income + hscore, 
  ziformula = ~ age,
  data = docvisit,
  family = poisson()
)
out <- check_overdispersion(mzip)
out
#> # Overdispersion test
#> 
#>  dispersion ratio =   1.417
#>           p-value = < 0.001
#> Overdispersion detected.
plot(out)
#> `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'
#> `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'
```

![](https://i.imgur.com/ayXuGoE.png)

``` r


mzinb <- glmmTMB(
  doctorco ~ sex + illness + income + hscore,
  ziformula = ~ age,
  data = docvisit,
  family = nbinom2()
)
out <- check_overdispersion(mzinb)
out
#> # Overdispersion test
#> 
#>  dispersion ratio = 1.031
#>           p-value =  0.64
#> No overdispersion detected.
plot(out)
#> `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'
#> `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'
```

![](https://i.imgur.com/BTVMw8U.png)

``` r

mzinbd <- glmmTMB(
  doctorco ~ sex + illness + income + hscore + age,
  ziformula = ~ sex + illness + income + hscore + age,
  dispformula = ~ sex + illness + income + hscore + age,
  data = docvisit,
  family = nbinom2()
)
out <- check_overdispersion(mzinbd)
out
#> # Overdispersion test
#> 
#>  dispersion ratio = 1.133
#>           p-value = 0.104
#> No overdispersion detected.
plot(out)
#> `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'
#> `geom_smooth()` using method = 'gam' and formula = 'y ~ s(x, bs = "cs")'
```

![](https://i.imgur.com/VKPdQ6S.png)

<sup>Created on 2024-03-17 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Looks like "nbinom2()" is currently inaccurate, the code we use is here:
https://github.com/easystats/performance/blob/35b5e19988386b584d91116be542baca1e98f33f/R/check_model_diagnostics.R#L370

(also pinging @bbolker and cross referencing to #654)

**Comment by strengejacke** [comment]
Closed in #913
